fendermint/vm/topdown/src/finality.rs

```diff
@@ -106,114 +195,64 @@ impl ParentViewProvider for InMemoryFinalityProvider {
 
         Ok(())
     }
-}
-
-impl ParentFinalityProvider for InMemoryFinalityProvider {
-    fn last_committed_finality(&self) -> StmResult<IPCParentFinality, Error> {
-        let finality = self
-            .last_committed_finality
-            .read_clone()?
-            .ok_or(StmError::Abort(Error::CommittedFinalityNotReady))?;
-        Ok(finality)
-    }
 
-    fn next_proposal(&self) -> StmResult<Option<IPCParentFinality>, Error> {
-        let latest_height = if let Some(h) = self.parent_view_data.latest_height()? {
-            h
+    fn check_height(&self, proposal: &IPCParentFinality) -> Stm<bool> {
+        let binding = self.last_committed_finality.read()?;
+        // last committed finality is not ready yet, we don't vote, just reject
+        let last_committed_finality = if let Some(f) = binding.as_ref() {
+            f
         } else {
-            return Ok(None);
+            return Ok(false);
         };
 
-        // latest height has not reached, we should wait or abort
-        if latest_height < self.config.chain_head_delay {
-            return Ok(None);
+        // the incoming proposal has height already committed, reject
+        if last_committed_finality.height >= proposal.height {
+            return Ok(false);
         }
 
-        let height = latest_height - self.config.chain_head_delay;
-        let last_committed_finality = self.last_committed_finality()?;
-
-        // parent height is not ready to be proposed yet
-        if height <= last_committed_finality.height {
-            return Ok(None);
-        }
-
-        // prepare block hash and validator set
-        let height_data = self.parent_view_data.height_data.read()?;
-        let block_hash = if let Some(v) = height_data.get_value(height) {
-            let (block_hash, _, _) = v;
-            block_hash.clone()
+        let latest_height = if let Some(h) = self.cached_data.latest_height()? {
+            h
         } else {
-            return abort(Error::HeightNotFoundInCache(height));
+            // latest height is not found, meaning we dont have any prefetched cache, we just be
+            // optimistic and vote yes as the previous latest committed finality check is passed
+            // and it's valid.
+            return Ok(true);
```

Comment:
I could be wrong but this doesn't compute to me. 

Looking at `check_block_hash`, a proposal _will_ be rejected if we don't know have data about that height, that is, we don't consider it final. For the same reason, we should reject this as well: we simply don't know! The "previous latest commmitted finality check" only says that the new proposal is higher than what we considered final last time, but that could be the tip of the chain, or something that doesn't even exist.

I don't think we should vote "yes" just because we know we will vote "no" in the other method, and we know the actual check method calls both. They should be self-contained IMO.

Reply:
I changed to `false`